src/lib/operators/join_hash.cpp

```diff
@@ -721,46 +218,83 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
     std::vector<std::vector<size_t>> histograms_left;
     std::vector<std::vector<size_t>> histograms_right;
-    /*
-    NUMA notes:
-    The materialized vectors don't have any strong NUMA preference because they haven't been partitioned yet.
-    However, it would be a good idea to keep each materialized vector on one node if possible.
-    This helps choosing a scheduler node for the radix phase (see below).
-    */
-    // Scheduler note: parallelize this at some point. Currently, the amount of jobs would be too high
-    auto materialized_left = materialize_input<LeftType, HashedType>(left_in_table, _column_ids.first, histograms_left,
-                                                                     _radix_bits);
-    // 'keep_nulls' makes sure that the relation on the right materializes NULL values when executing an OUTER join.
-    auto materialized_right = materialize_input<RightType, HashedType>(
-        right_in_table, _column_ids.second, histograms_right, _radix_bits, keep_nulls);
-
-    // Radix Partitioning phase
-    /*
-    NUMA notes:
-    If the input vectors (the materialized vectors) reside on a specific node, the partitioning worker for
-    this phase should be scheduled on the same node.
-    Additionally, the output vectors in this phase are partitioned by a radix key. Therefore it would be good
-    to pin the outputs from both sides on the same node for each radix partition. For example, if there are
-    only two radix partitions A and B, the partitions leftA and rightA should be on the same node, and the
-    partitions leftB and leftB should also be on the same node.
-    */
-    // Scheduler note: parallelize this at some point. Currently, the amount of jobs would be too high
-    auto radix_left =
-        partition_radix_parallel<LeftType>(materialized_left, left_chunk_offsets, histograms_left, _radix_bits);
-    // 'keep_nulls' makes sure that the relation on the right keeps NULL values when executing an OUTER join.
-    auto radix_right = partition_radix_parallel<RightType>(materialized_right, right_chunk_offsets, histograms_right,
-                                                           _radix_bits, keep_nulls);
 
-    // Build phase
-    auto hashtables = build<LeftType, HashedType>(radix_left);
+    // Output containers of materialization phase. Type similar to the output
+    // of radix partitioning phase to allow short cut for _radix_bits == 0
+    // (in this case, we can skip the partitioning alltogether).
+    RadixContainer<LeftType> materialized_left;
+    RadixContainer<RightType> materialized_right;
+
+    // Containers for potential (skipped when left side small) radix partitioning phase
+    RadixContainer<LeftType> radix_left;
+    RadixContainer<RightType> radix_right;
+    std::vector<std::optional<HashTable<HashedType>>> hashtables;
+
+    // Parallelization of Hash Join:
+    // we have to data paths, one for left side and one for right side:
```

Comment:
Where do I see anything about parallelization here?

Reply:
Ah, I see. That only makes sense when you look at the following code.

Maybe add a comment that the left and right side are prepared in parallel